src/Tracepoints.jl

```diff
@@ -0,0 +1,497 @@
+    end
+end
+function create_tracepoint!(mod::Module, source, category::Symbol, kind::Symbol, lib_id, args::Expr; alias_with=nothing)
+    argspec = parse_args(mod, args)
+    spec = create_tracepoint_spec(source, argspec, args; alias_with)
+    register_probe(mod, category, kind, spec)
+    return create_tracepoint_call(category, kind, lib_id, argspec, spec)
+end
+# Semaphore and probe payload not yet loaded
+function create_tracepoint_call(category, kind, lib_id, argspec, spec)
+    abi_type, args_box, arg_name = wrap_tracepoint_args(argspec)
+    return quote
+        $probe_maybe_trigger($spec, $(QuoteNode(category)), $(QuoteNode(kind)), $lib_id, $(Val{abi_type}()), $args_box, $(QuoteNode(arg_name)))
+    end
+end
+# Semaphore and probe payload already loaded
+function create_tracepoint_call_loaded(category, kind, lib_id, argspec, spec, sema_set, payload)
+    abi_type, args_box, arg_name = wrap_tracepoint_args(argspec)
+    return quote
+        if $sema_set && $payload != C_NULL
+            $probe_trigger($spec, $(QuoteNode(category)), $(QuoteNode(kind)), $lib_id, $(Val{abi_type}()), $args_box, $(QuoteNode(arg_name)))
+        end
+    end
+end
+
+macro tracepoint(kind::QuoteNode, category::QuoteNode, lib_id, args=Expr(:tuple))
+    create_tracepoint!(__module__, __source__, category.value, kind.value, lib_id, args)
+end
+macro region_start(category::QuoteNode, args=Expr(:tuple))
+    create_tracepoint!(__module__, __source__, category.value, :start, 0, args)
+end
+macro region_finish(category::QuoteNode, lib_id, args=Expr(:tuple))
+    create_tracepoint!(__module__, __source__, category.value, :finish, lib_id, args)
+end
+
+"""
+    @region :category arg1 arg2... ex
+
+Constructs a pair of tracepoints around the execution of `ex`. The start
+tracepoint executes just before `ex`, and the finish tracepoint executes just
+after `ex`. The category for both tracepoints is `:category`, and must be a
+literal `Symbol` (as the tracepoints are constructed while `@region` is being
+parsed). The kind for each tracepoint is `:start` and `:finish` respectively
+(used when programming or enabling/disabling tracepoints).
+
+These tracepoints start out disabled and unprogrammed, and by default will do
+nothing (having negligible performance impact, if any). They can be programmed
+with [`set!`](@ref), and enabled with [`enable!`](@ref).
+
+Arguments may be provided in a few formats:
+
+`arg::T` - This specifies that the argument with name `arg` has type `T`, where
+`T` is a concrete type. In this form, `arg` must be a variable in scope that
+`@region` can read during start and finish tracepoint execution. If `arg` is
+modified during the execution of `ex`, its new value will be passed into the
+finish tracepoint.
+
+`arg::T=value` - This format is similar to `arg::T`, except that `arg` is not
+read from a variable with the same name, but is instead hard-coded as `value`
+for both start and finish tracepoints.
```

Comment:
```suggestion
read from a variable with the same name, but is instead evaluated at parse-time
and the result is hard-coded for both start and finish tracepoints.
```

I find the different evaluation time for this syntax confusing. Why not just have it evaluate at runtime the same way the `arg::T` syntax does?

It seems like there's no way, e.g., to pass a dynamically-constructed tuple `(x,y)` to the API without copying it into a local with the intended name.

Reply:
It's not evaluated at parse-time - I admit `hard-coded` is not what I meant, it's really just interpolated at runtime as you'd expect (if this is not true, please let me know!).